zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2001,44 +2008,65 @@ protected void doRetry() {
     final long startT = System.currentTimeMillis();
-    retryUntilConnected(() -> {
-      ((ZkConnection) getConnection()).getZookeeper().getData(path, null, cb,
-          new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, true) {
-            @Override
-            protected void doRetry() {
-              asyncGetData(path, cb);
-            }
-          });
-      return null;
-    });
+    try {
+      retryUntilConnected(() -> {
+        ((ZkConnection) getConnection()).getZookeeper().getData(path, null, cb,
+            new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, true) {
+              @Override
+              protected void doRetry() {
+                asyncGetData(path, cb);
+              }
+            });
+        return null;
+      });
+    } catch (RuntimeException e) {
+      // Process callback to release caller from waiting
+      cb.processResult(ZkAsyncCallbacks.UNKNOWN_RET_CODE, path,
+          new ZkAsyncCallMonitorContext(_monitor, startT, 0, true), null, null);
+      throw e;
+    }
   }
 
   public void asyncExists(final String path, final ZkAsyncCallbacks.ExistsCallbackHandler cb) {
     final long startT = System.currentTimeMillis();
-    retryUntilConnected(() -> {
-      ((ZkConnection) getConnection()).getZookeeper().exists(path, null, cb,
-          new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, true) {
-            @Override
-            protected void doRetry() {
-              asyncExists(path, cb);
-            }
-          });
-      return null;
-    });
+    try {
+      retryUntilConnected(() -> {
+        ((ZkConnection) getConnection()).getZookeeper().exists(path, null, cb,
+            new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, true) {
+              @Override
+              protected void doRetry() {
+                asyncExists(path, cb);
+              }
+            });
+        return null;
+      });
+    } catch (RuntimeException e) {
+      // Process callback to release caller from waiting
+      cb.processResult(ZkAsyncCallbacks.UNKNOWN_RET_CODE, path,
+          new ZkAsyncCallMonitorContext(_monitor, startT, 0, true), null);
+      throw e;
+    }
   }
 
   public void asyncDelete(final String path, final ZkAsyncCallbacks.DeleteCallbackHandler cb) {
     final long startT = System.currentTimeMillis();
-    retryUntilConnected(() -> {
-      ((ZkConnection) getConnection()).getZookeeper().delete(path, -1, cb,
-          new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, false) {
-            @Override
-            protected void doRetry() {
-              asyncDelete(path, cb);
-            }
-          });
-      return null;
-    });
+    try {
+      retryUntilConnected(() -> {
+        ((ZkConnection) getConnection()).getZookeeper().delete(path, -1, cb,
```

Comment:
I believe the only known case that this could fail on the first try is that if we use "getExpectedZookeeper(expectedSessionId)" to avoid the unexpected updates after a session change. Given that says,
1. @rabashizade did we see any real concern that it may fail on the first try? Or you are just trying to add more protection? I think the change is good, just want to understand if any real case you foresee.
2. @pkuwm any idea why we are not using getExpectedZookeeper(expectedSessionId) for the asyncDelete call?

Reply:
I didn't encounter any real issues, I mostly put the check there due to the types of exceptions that could be thrown by `retryUntilConnected()`, and so I wanted to be able to record the failure in the event that one happens.